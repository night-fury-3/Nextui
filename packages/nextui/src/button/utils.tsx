import React from 'react';
import { ButtonProps } from '../button/button';
import { ButtonGroupConfig } from '../button/button-group-context';

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    ghost: config.ghost || props.ghost,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
