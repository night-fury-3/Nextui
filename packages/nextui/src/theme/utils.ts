import { NextUIThemes } from './index';
import useWarning from '@hooks/use-warning';
import { DeepPartial } from '@utils/types';
import { deepMergeObject } from '@utils/object';
import darkTheme from './dark';
import lightTheme from './default';

export type PartialTheme = DeepPartial<NextUIThemes>;

export type ThemeParam =
  | PartialTheme
  | ((theme: PartialTheme) => PartialTheme)
  | undefined;

export const mergeTheme = (
  current: NextUIThemes,
  custom: ThemeParam
): NextUIThemes => {
  if (!custom) return current;
  if (typeof custom === 'function') {
    const merged = custom(current);
    if (!merged || typeof merged !== 'object') {
      useWarning('The theme function must return object value.');
    }
    return merged as NextUIThemes;
  }
  return deepMergeObject<NextUIThemes>(current, custom as NextUIThemes);
};

export const switchTheme = (mergedTheme: NextUIThemes): NextUIThemes => {
  const themes: { [key in NextUIThemes['type']]: NextUIThemes } = {
    light: lightTheme,
    dark: darkTheme,
  };
  return { ...mergedTheme, ...themes[mergedTheme.type] };
};
