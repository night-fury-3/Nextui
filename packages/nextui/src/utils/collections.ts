import React, { ReactNode } from 'react';

export const getId = () => {
  return Math.random().toString(32).slice(2, 10);
};

export const hasChild = (
  children: ReactNode | undefined,
  child: React.ElementType
): boolean => {
  const types = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return null;
    return item.type;
  });

  return (types || []).includes(child);
};

export const pick = <Obj extends { [key: string]: any }, Key extends keyof Obj>(
  props: Key[],
  obj: Obj
): Pick<Obj, Key> =>
  props.reduce((acc, prop) => {
    acc[prop] = obj[prop];
    return acc;
  }, {} as Pick<Obj, Key>);

export const pickChild = (
  children: ReactNode | undefined,
  targetChild: React.ElementType
): [ReactNode | undefined, ReactNode | undefined] => {
  let target: ReactNode[] = [];
  const withoutTargetChildren = React.Children.map(children, (item) => {
    if (!React.isValidElement(item)) return item;
    if (item.type === targetChild) {
      target.push(item);
      return null;
    }
    return item;
  });

  const targetChildren = target.length >= 0 ? target : undefined;

  return [withoutTargetChildren, targetChildren];
};
