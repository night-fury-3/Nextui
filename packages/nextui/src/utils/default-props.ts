import { NormalSizes } from './prop-types';
import type { CSSProperties } from 'react';

export type NextUISpacingWithAuto = NormalSizes | 'px' | 'auto' | number;

export interface NextUIMargins {
  m?: NextUISpacingWithAuto;
  my?: NextUISpacingWithAuto;
  mx?: NextUISpacingWithAuto;
  mt?: NextUISpacingWithAuto;
  mb?: NextUISpacingWithAuto;
  ml?: NextUISpacingWithAuto;
  mr?: NextUISpacingWithAuto;
}

export interface NextUIPaddings {
  p?: NextUISpacingWithAuto;
  py?: NextUISpacingWithAuto;
  px?: NextUISpacingWithAuto;
  pt?: NextUISpacingWithAuto;
  pb?: NextUISpacingWithAuto;
  pl?: NextUISpacingWithAuto;
  pr?: NextUISpacingWithAuto;
}

export type NextUISpaces = NextUIMargins & NextUIPaddings;

export type NextUISpacesKeys = { [key in keyof NextUISpaces]?: string };

export const NextUISpacesStringKeys = [
  'm',
  'my',
  'mx',
  'mt',
  'mb',
  'ml',
  'mr',
  'p',
  'py',
  'px',
  'pt',
  'pb',
  'pl',
  'pr'
];

export interface DefaultProps extends NextUIMargins, NextUIPaddings {
  style?: CSSProperties;
  className?: string;
}
