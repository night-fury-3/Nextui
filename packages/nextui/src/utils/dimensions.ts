import { NormalWeights, NormalSizes } from './prop-types';

interface GetSizeValue {
  size: string | number;
  sizes: Record<string, any>;
  defaultSize?: string;
}

export const getMargin = (num: number): string => {
  return `calc(${num * 15.25}pt + 1px * ${num - 1})`;
};

export const getNormalWeight = (weight?: NormalWeights): string | undefined => {
  const weights: { [key in NormalWeights]?: string } = {
    light: '1px',
    normal: '2px',
    bold: '3px'
  };
  return weights[weight || 'normal'];
};

export const getNormalRadius = (
  size: NormalSizes,
  rounded?: boolean
): string => {
  const radius: { [key in NormalSizes]: string } = {
    mini: '7px',
    small: '9px',
    medium: '12px',
    large: '14px',
    xlarge: '15px'
  };
  const baseRadius = radius[size];
  return rounded ? `calc(${baseRadius} + 10rem)` : baseRadius;
};

export function getSizeValue({
  size,
  sizes,
  defaultSize = 'md'
}: GetSizeValue) {
  if (typeof size === 'number') {
    return size;
  }
  return sizes[size] || size || sizes[defaultSize];
}
