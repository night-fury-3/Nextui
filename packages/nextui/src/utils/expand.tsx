import React, { useEffect, useRef, useState } from 'react';
import withDefaults from '../utils/with-defaults';
import useRealShape from '../use-real-shape';
import clsx from './clsx';

interface Props {
  isExpanded?: boolean;
  animated?: boolean;
  delay?: number;
}

const defaultProps = {
  isExpanded: false,
  animated: true,
  delay: 200
};

export type ExpandProps = Props & typeof defaultProps;

const Expand: React.FC<React.PropsWithChildren<ExpandProps>> = ({
  isExpanded,
  delay,
  animated,
  children
}) => {
  const [height, setHeight] = useState<string>(isExpanded ? 'auto' : '0');
  const [selfExpanded, setSelfExpanded] = useState<boolean>(isExpanded);
  const contentRef = useRef<HTMLDivElement>(null);
  const entryTimer = useRef<number>();
  const leaveTimer = useRef<number>();
  const resetTimer = useRef<number>();
  const [state, updateShape] = useRealShape<HTMLDivElement>(contentRef);

  useEffect(() => setHeight(`${state.height}px`), [state.height]);
  useEffect(() => {
    // show element or reset height.
    // force an update once manually, even if the element does not change.
    // (the height of the element might be "auto")
    if (!isExpanded) {
      updateShape();
      setHeight(`${state.height}px`);
    }

    // show expand animation
    entryTimer.current = window.setTimeout(() => {
      setSelfExpanded(isExpanded);
      clearTimeout(entryTimer.current);
    }, 30);

    // Reset height after animation
    if (isExpanded) {
      resetTimer.current = window.setTimeout(() => {
        setHeight('auto');
        clearTimeout(resetTimer.current);
      }, delay);
    } else {
      leaveTimer.current = window.setTimeout(() => {
        clearTimeout(leaveTimer.current);
      }, delay / 2);
    }

    return () => {
      clearTimeout(entryTimer.current);
      clearTimeout(leaveTimer.current);
      clearTimeout(resetTimer.current);
    };
  }, [isExpanded]);

  return (
    <div className={clsx('container', { expanded: selfExpanded, animated })}>
      <div ref={contentRef} className="content">
        {children}
      </div>
      <style jsx>{`
        .container {
          padding: 0;
          margin: 0;
          height: 0;
          opacity: 0;
          overflow: hidden;
        }
        .animated {
          transition: height ${delay}ms ease 0ms,
            opacity ${delay * 1.5}ms ease 0ms;
        }
        .expanded {
          height: ${height};
          opacity: 1;
        }
      `}</style>
    </div>
  );
};

export default withDefaults(Expand, defaultProps);
