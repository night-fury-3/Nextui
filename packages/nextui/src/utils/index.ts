import { NormalColors, normalColors } from './prop-types';
import { NextUIThemesPalette } from '@theme/index';
/**
 * This function allows validate if a string is a hexadecimal
 * value
 * @param str [string] hexadecimal value
 * @returns result [boolean]
 */
export const isHex = (str: string): boolean => {
  const exp = /#[a-fA-F0-9]{3,6}/g;
  return exp.test(str);
};

/**
 *
 * @param hex [string]
 * @param alpha [number]
 * @returns [string]
 */
export const hexToRGBA = (hex: string, alpha: number = 1): string => {
  let r: string | number = 0,
    g: string | number = 0,
    b: string | number = 0;
  // 3 digits
  if (hex.length == 4) {
    r = '0x' + hex[1] + hex[1];
    g = '0x' + hex[2] + hex[2];
    b = '0x' + hex[3] + hex[3];
    // 6 digits
  } else if (hex.length == 7) {
    r = '0x' + hex[1] + hex[2];
    g = '0x' + hex[3] + hex[4];
    b = '0x' + hex[5] + hex[6];
  }
  return `rgba(${+r}, ${+g},${+b},${alpha})`;
};

export const getNormalColor = (
  color: NormalColors | string | undefined,
  palette: NextUIThemesPalette,
  defaultColor: string = 'inherit'
) => {
  const colors: { [key in NormalColors | string]: string } = {
    default: defaultColor,
    primary: palette.primary,
    secondary: palette.secondary,
    success: palette.success,
    warning: palette.warning,
    error: palette.error,
    gradient: palette.gradient,
  };
  return color && colors[color] ? colors[color] : defaultColor;
};

export const isNormalColor = (color: string): boolean => {
  let found = normalColors.find((el) => el === color);
  return found !== undefined && found !== null;
};

/**
 * Function that checks color name support in the current browser
 * @param strColor
 * @returns boolean
 */
export const isColor = (strColor: string) => {
  let s = new Option().style;
  s.color = strColor;
  return s.color == strColor;
};
