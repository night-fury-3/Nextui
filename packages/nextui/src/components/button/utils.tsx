import React, { ReactNode } from 'react';
import { NormalSizes } from '@utils/prop-types';
import ButtonIcon from './button-icon';
import { ButtonProps } from '@components/button/button';
import { ButtonGroupConfig } from '@components/button/button-group-context';

export const getButtonChildrenWithIcon = (
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode;
    iconRight?: React.ReactNode;
  }
) => {
  const { icon, iconRight } = icons;
  const hasIcon = icon || iconRight;
  const isRight = Boolean(iconRight);
  const paddingForAutoMode =
    auto || size === 'mini'
      ? `calc(var(--next-ui-button-height) / 2 + var(--next-ui-button-padding) * .5)`
      : 0;
  if (!hasIcon) return <div className="text">{children}</div>;
  if (React.Children.count(children) === 0) {
    return (
      <ButtonIcon isRight={isRight} isSingle>
        {hasIcon}
      </ButtonIcon>
    );
  }
  return (
    <React.Fragment>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </React.Fragment>
  );
};

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig
): ButtonProps => {
  if (!config.isButtonGroup) return props;
  return {
    ...props,
    auto: true,
    shadow: false,
    bordered: config.bordered || props.bordered,
    flat: config.flat || props.flat,
    animated: config.animated || props.animated,
    rounded: config.rounded || props.rounded,
    light: config.light || props.light,
    size: config.size || props.size,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  };
};
