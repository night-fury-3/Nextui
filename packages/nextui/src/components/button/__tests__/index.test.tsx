import React from 'react';
import { mount } from 'enzyme';
import Button from '../index';
import { sleep } from '../../../../tests/utils';

describe('Button', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Button>Button</Button>);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should support all colors', () => {
    const wrapper = mount(
      <div>
        <Button color="primary" />
        <Button color="secondary" />
        <Button color="success" />
        <Button color="warning" />
        <Button color="error" />
        <Button color="gradient" />
      </div>
    );
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should support all sizes', () => {
    const wrapper = mount(
      <div>
        <Button size="mini" />
        <Button size="small" />
        <Button size="medium" />
        <Button size="large" />
        <Button size="xlarge" />
      </div>
    );
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should render different text', () => {
    const wrapper = mount(<Button>button</Button>);
    expect(wrapper.text()).toContain('button');

    wrapper.setProps({
      children: <span>Hello</span>,
    });
    expect(wrapper.text()).toContain('Hello');
  });

  it('should render empty button correctly', () => {
    expect(<Button />).toMatchSnapshot();
  });

  it('should trigger callback function', () => {
    const WrapperButton = () => {
      const [state, setState] = React.useState<string>('state1');
      return <Button onClick={() => setState('state2')}>{state}</Button>;
    };
    const wrapper = mount(<WrapperButton />);
    expect(wrapper.text()).toContain('state1');

    wrapper.simulate('click');
    expect(wrapper.text()).toContain('state2');
  });

  it('should ignore events when disabled', () => {
    const WrapperButton = () => {
      const [state, setState] = React.useState<string>('state1');
      return (
        <Button disabled onClick={() => setState('state2')}>
          {state}
        </Button>
      );
    };
    const wrapper = mount(<WrapperButton />);
    expect(wrapper.text()).toContain('state1');

    wrapper.simulate('click');
    expect(wrapper.text()).toContain('state1');
    expect(wrapper.text()).not.toContain('state2');
  });

  it('should ignore events when loading', () => {
    const WrapperButton = () => {
      const [state, setState] = React.useState<string>('state1');
      return (
        <Button loading onClick={() => setState('state2')}>
          {state}
        </Button>
      );
    };
    const wrapper = mount(<WrapperButton />);
    wrapper.simulate('click');
    expect(wrapper.text()).not.toContain('state2');
  });

  it('should render different variants', () => {
    const wrapper = mount(
      <div>
        <Button flat>button</Button>
        <Button light color="warning">
          light
        </Button>
        <Button flat color="success">
          button
        </Button>
        <Button flat color="warning">
          button
        </Button>
        <Button rounded loading>
          button
        </Button>
        <Button flat loading>
          button
        </Button>
        <Button shadow>button</Button>
        {/* <Button ghost>button</Button>
        <Button bordered>button</Button> */}
        <Button auto>button</Button>
        <Button animated={false}>button</Button>
      </div>
    );
    expect(wrapper).toMatchSnapshot();
    expect(<Button loading>button</Button>).toMatchSnapshot();
  });

  it('should remove expired events', () => {
    const wrapper = mount(<Button>button</Button>);
    wrapper.simulate('click');
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it('should support loading change with deply', async () => {
    const wrapper = mount(<Button>button</Button>);
    wrapper.simulate('click');
    await sleep(500);
    wrapper.setProps({ loading: true });
    await sleep(500);
    expect(wrapper.find('.loading-container').length).not.toBe(0);
  });

  it('ref should be forwarded', () => {
    const ref = React.createRef<HTMLButtonElement>();
    const wrapper = mount(<Button ref={ref}>action</Button>);
    expect(wrapper.find('button').getDOMNode()).toEqual(ref.current);
    expect(() => wrapper.unmount()).not.toThrow();
  });
});
