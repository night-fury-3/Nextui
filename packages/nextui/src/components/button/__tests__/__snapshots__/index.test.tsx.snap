// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render different variants 1`] = `ReactWrapper {}`;

exports[`Button should render different variants 2`] = `
<NextUI - Button
  animated={true}
  auto={false}
  bordered={false}
  className=""
  color="default"
  disabled={false}
  flat={false}
  htmlType="button"
  light={false}
  loaderType="default"
  loading={true}
  rounded={false}
  shadow={false}
  size="medium"
  weight="normal"
>
  button
</NextUI - Button>
`;

exports[`Button should render empty button correctly 1`] = `
<NextUI - Button
  animated={true}
  auto={false}
  bordered={false}
  className=""
  color="default"
  disabled={false}
  flat={false}
  htmlType="button"
  light={false}
  loaderType="default"
  loading={false}
  rounded={false}
  shadow={false}
  size="medium"
  weight="normal"
/>
`;
